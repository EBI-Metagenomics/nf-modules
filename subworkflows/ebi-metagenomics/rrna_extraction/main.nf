
include { INFERNAL_CMSEARCH           } from '../../../modules/ebi-metagenomics/infernal/cmsearch/main'
include { CMSEARCHTBLOUTDEOVERLAP     } from '../../../modules/ebi-metagenomics/cmsearchtbloutdeoverlap/main'
include { EASEL_ESLSFETCH             } from '../../../modules/ebi-metagenomics/easel/eslsfetch/main'
include { EXTRACTCOORDS               } from '../../../modules/ebi-metagenomics/extractcoords/main'

workflow RRNA_EXTRACTION {

    take:
    ch_fasta     // channel: [ val(meta), [ fasta ] ]
    rfam         // file: rfam for cmsearch
    claninfo     // file: claninfo for cmsearchtbloutdeoverlap

    main:

    ch_versions = Channel.empty()

    INFERNAL_CMSEARCH(
        ch_fasta,
        rfam
    )
    ch_versions = ch_versions.mix(INFERNAL_CMSEARCH.out.versions.first())

    CMSEARCHTBLOUTDEOVERLAP(
        INFERNAL_CMSEARCH.out.cmsearch_tbl,
        claninfo
    )
    ch_versions = ch_versions.mix(CMSEARCHTBLOUTDEOVERLAP.out.versions.first())

    ch_easel = ch_fasta
                .join(CMSEARCHTBLOUTDEOVERLAP.out.cmsearch_tblout_deoverlapped)

    EASEL_ESLSFETCH(
        ch_easel
    )
    ch_versions = ch_versions.mix(EASEL_ESLSFETCH.out.versions.first())

    EXTRACTCOORDS(
        EASEL_ESLSFETCH.out.easel_coords,
        EASEL_ESLSFETCH.out.matched_seqs_with_coords,
        false
    )
    ch_versions = ch_versions.mix(EXTRACTCOORDS.out.versions.first())

    emit:
    cmsearch_deoverlap_out = CMSEARCHTBLOUTDEOVERLAP.out.cmsearch_tblout_deoverlapped   // channel: [ val(meta), [ deoverlapped ] ]
    easel_out = EASEL_ESLSFETCH.out.easel_coords                                        // channel: [ val(meta), [ fasta ] ]
    ssu_fasta = EXTRACTCOORDS.out.ssu_fasta                                             // channel: [ val(meta), [ fasta ] ]
    lsu_fasta = EXTRACTCOORDS.out.lsu_fasta                                             // channel: [ val(meta), [ fasta ] ]
    concat_ssu_lsu_coords = EXTRACTCOORDS.out.concat_ssu_lsu_coords                     // channel: [ val(meta), [ txt ] ]
    versions = ch_versions                                                              // channel: [ versions.yml ]
}

