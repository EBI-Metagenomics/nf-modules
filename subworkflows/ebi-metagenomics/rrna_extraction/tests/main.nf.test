nextflow_workflow {

    name "Test Subworkflow RRNA_EXTRACTION"
    script "../main.nf"
    workflow "RRNA_EXTRACTION"
    config "./nextflow.config"

    tag "subworkflows"
    tag "subworkflows_ebimetagenomics"
    tag "subworkflows/rrna_extraction"
    tag "infernal"
    tag "infernal/cmsearch"
    tag "cmsearchtbloutdeoverlap"
    tag "easel"
    tag "easel/eslsfetch"
    tag "extractcoords"

    test("test_rrna_extraction") {

        when {
            params {
                rfam = "${moduleDir}/tests/"
                rfam_clan = "${moduleDir}/tests/ribo.claninfo"
            }
            workflow {
                """
                input[0] = Channel.of( [ [ id:'test', single_end:false ], // meta map
                    file('${moduleDir}/tests/test.fasta.gz', checkIfExists: true)
                ])
                input[1] = file(params.rfam)
                input[2] = file(params.rfam_clan)
                input[3] = false
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
