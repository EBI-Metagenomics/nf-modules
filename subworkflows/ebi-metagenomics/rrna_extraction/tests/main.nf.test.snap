{
    "test_rrna_extraction": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.tblout.deoverlapped:md5,2ff422ba8581127bc95262d9751dbcba"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_easel_coords.fasta:md5,fb23f13526672a300845b767dbc5fd44"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_SSU.fasta:md5,fb23f13526672a300845b767dbc5fd44"
                    ]
                ],
                "3": [
                    
                ],
                "4": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_concat_SSU_LSU_coords.txt:md5,401222d263116bb145c1a67e6acc4709"
                    ]
                ],
                "5": [
                    "versions.yml:md5,1cd8a1b942aaba58bd5d6ab9c4f3f9b2",
                    "versions.yml:md5,986b50b86874f29db48120870f7ac4f6",
                    "versions.yml:md5,a83addcfaf303617fdda5c6a5958a5ad",
                    "versions.yml:md5,d1294ff919ea50bc3a41ddc5a88ab3de"
                ],
                "cmsearch_deoverlap_out": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.tblout.deoverlapped:md5,2ff422ba8581127bc95262d9751dbcba"
                    ]
                ],
                "concat_ssu_lsu_coords": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_concat_SSU_LSU_coords.txt:md5,401222d263116bb145c1a67e6acc4709"
                    ]
                ],
                "easel_out": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_easel_coords.fasta:md5,fb23f13526672a300845b767dbc5fd44"
                    ]
                ],
                "lsu_fasta": [
                    
                ],
                "ssu_fasta": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_SSU.fasta:md5,fb23f13526672a300845b767dbc5fd44"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,1cd8a1b942aaba58bd5d6ab9c4f3f9b2",
                    "versions.yml:md5,986b50b86874f29db48120870f7ac4f6",
                    "versions.yml:md5,a83addcfaf303617fdda5c6a5958a5ad",
                    "versions.yml:md5,d1294ff919ea50bc3a41ddc5a88ab3de"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-06T09:16:39.686196"
    }
}